scroll_derive/examples/derive_custom_ctx.rs

```diff
@@ -0,0 +1,81 @@
+use scroll_derive::{Pread, Pwrite, SizeWith};
+
+/// An example of using a method as the value for a ctx in a derive.
+struct EndianDependent(Endian);
+impl EndianDependent {
+    fn len(&self) -> usize {
+        match self.0 {
+            scroll::Endian::Little => 5,
+            scroll::Endian::Big => 6,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+struct VariableLengthData {
+    buf: Vec<u8>,
+}
+
+impl<'a> TryFromCtx<'a, usize> for VariableLengthData {
+    type Error = scroll::Error;
+
+    fn try_from_ctx(from: &'a [u8], ctx: usize) -> Result<(Self, usize), Self::Error> {
+        let offset = &mut 0;
+        let buf = from.gread_with::<&[u8]>(offset, ctx)?.to_owned();
+        Ok((Self { buf }, *offset))
+    }
+}
+impl<'a> TryIntoCtx<usize> for &'a VariableLengthData {
+    type Error = scroll::Error;
+    fn try_into_ctx(self, dst: &mut [u8], ctx: usize) -> Result<usize, Self::Error> {
+        let offset = &mut 0;
+        for i in 0..(ctx.min(self.buf.len())) {
+            dst.gwrite(self.buf[i], offset)?;
+        }
+        Ok(*offset)
+    }
+}
+impl SizeWith<usize> for VariableLengthData {
+    fn size_with(ctx: &usize) -> usize {
+        *ctx
+    }
+}
+
+#[derive(Debug, PartialEq, Pread, Pwrite, SizeWith)]
+#[repr(C)]
+struct Data {
+    id: u32,
+    timestamp: f64,
+    // You can fix the ctx regardless of what is passed in.
+    #[scroll(ctx = BE)]
+    arr: [u16; 2],
+    // You can use arbitrary expressions for the ctx.
+    // You have access to the `ctx` parameter of the `{pread/gread}_with` inside the expression.
```

Comment:
this is so wild hah, i have to think about usecases/what it means, or if there's some kind of hygiene issue that a different design might not have, but it's very interesting! is the fact that `ctx` is available within the scope of the pread something you have made available or is it a property of derive with field overrides in macros?

Reply:
It's because the user's expression is effectively copy-pasted without hygiene and so the parameters of the trait method are accessible. It would take more work to remove this ability, since we'd have to parse for an instance of a `ctx` ident and actively block it.

Reply:
This also means that if only `Pwrite` derive is used, they won't get an error using `self` or `dst` idents.